xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/objects/classes/DBListClass.java

```diff
@@ -294,8 +305,17 @@ public String getQuery(XWikiContext context)
                         select.append(", doc." + valueField);
                     } else {
                         select.append(", valueprop.value");
-                        fromStatements.add("StringProperty as valueprop");
                         whereStatements.add("obj.id=valueprop.id.id and valueprop.id.name='" + valueField + "'");
+                        // Get the from statements according to the type of property
+                        try {
+                            PropertyClass pc = (PropertyClass) context.getWiki().getDocument(classname,
+                                    context).getXClass()
+                                    .get(valueField);
+                            String classType = (pc == null) ? "StringProperty" : pc.newProperty().getClass().getName();
+                            fromStatements.add(classType + "  as valueprop");
+                        } catch (XWikiException e) {
+                            e.printStackTrace();
```

Comment:
BTW (not related to your commit), it seems we have SQ-injection in this method, we need to fix this too with parametrized queries.